crates/sui-framework/sources/collectible/transfer_policy.move

```diff
@@ -0,0 +1,448 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Part of the collectibles bundle - a primitive allowing creators to enforce
+/// constraints on transfers as long as the transfers are performed in the ecosystem
+/// that enforces them.
+module sui::transfer_policy {
+    use std::vector;
+    use std::option::{Self, Option};
+    use std::type_name::{Self, TypeName};
+    use sui::package::{Self, Publisher};
+    use sui::tx_context::TxContext;
+    use sui::object::{Self, ID, UID};
+    use sui::vec_set::{Self, VecSet};
+    use sui::dynamic_field as df;
+    use sui::bag::{Self, Bag};
+    use sui::balance::{Self, Balance};
+    use sui::sui::SUI;
+    use sui::coin::{Self, Coin};
+    use sui::event;
+
+    friend sui::collectible;
+
+    /// The number of receipts does not match the `TransferPolicy` requirement.
+    const EPolicyNotSatisfied: u64 = 0;
+    /// A completed rule is not set in the `TransferPolicy`.
+    const EIllegalRule: u64 = 1;
+    /// A Rule is not set.
+    const EUnknownRequrement: u64 = 2;
+    /// Attempting to create a Rule that is already set.
+    const ERuleAlreadySet: u64 = 3;
+    /// Trying to `withdraw` or `close_and_withdraw` with a wrong Cap.
+    const ENotOwner: u64 = 4;
+    /// Trying to `withdraw` more than there is.
+    const ENotEnough: u64 = 5;
+
+    /// A "Hot Potato" forcing the buyer to get a transfer permission
+    /// from the item type (`T`) owner on purchase attempt.
+    struct TransferRequest<phantom T: key + store> {
+        /// Amount of SUI paid for the item. Can be used to
+        /// calculate the fee / transfer policy enforcement.
+        paid: u64,
+        /// The ID of the Kiosk / Safe the object is being sold from.
+        /// Can be used by the TransferPolicy implementors.
+        from: ID,
+        /// A Bag of custom details attached to the `TransferRequest`.
+        /// The attachments must be resolved before the `TransferRequest`
+        /// can be completed and unpacked to accept the transfer.
+        metadata: Bag,
+        /// Collected Receipts. Used to verify that all of the rules
+        /// were followed and `TransferRequest` can be confirmed.
+        receipts: VecSet<TypeName>
```

Comment:
I had originally imagined that the bag of metadata would operate as receipts in reverse (i.e. rather than adding a receipt for every successfully processed rule, we remove its metadata). Just noting that idea down here, but I may learn why that isn't possible below.

Reply:
It's a mix. For the most common application there's no metadata, and you can see it in the test policies. However, if someone builds an extension for the wallet which adds some extra data + creates a special policy - then metadata is used. This thing is still in research but I believe OB may find it useful. 

Next goal is writing extensions for the kiosk.

Reply:
My question was more whether you needed both `receipts` and `metadata`?  I.e. if you had a policy that only needed a receipt, and not metadata, could it be implemented, using a marker type for the metadata that gets removed when the policy's criteria is met?

Reply:
Metadata only works if you assigned it previously while receipts can exist independently from the module where request was created. The flow here is the following:

1. Marketplace: new request (does not know anything about policies)
2. TransferPolicy: requires X, Y and Z 
3. Collect receipts from X, Y and Z and complete transfer

While with metadata it's like this:

1. Marketplace: new request + custom metadata
2. TransferPolicy: requires X, Y and Z
3. Collect receipts AND deal with metadata
4. Complete transfer

So these "constraints" or "obligations" come from different sources: metadata comes from the source module which creates a request and then requires an extra action which probably should be included somewhere (?) not only in policies; while receipts count and types are defined purely by the policy.

I know it might look tricky but it should become more clear once we have more examples of extensions built on top of the Kiosk or TransferPolicy.